server/pulp/server/controllers/repository.py

```diff
@@ -125,3 +145,708 @@ def associate_single_unit(repository, unit):
+    group_manager = manager_factory.repo_group_manager()
+    group_manager.remove_repo_from_groups(repo_id)
+
+    if len(error_tuples) > 0:
+        pe = pulp_exceptions.PulpExecutionException()
+        pe.child_exceptions = error_tuples
+        raise pe
+
+    # append unbind itineraries foreach bound consumer
+    options = {}
+    consumer_bind_manager = manager_factory.consumer_bind_manager()
+
+    additional_tasks = []
+    errors = []
+    for bind in consumer_bind_manager.find_by_repo(repo_id):
+        try:
+            report = consumer_controller.unbind(bind['consumer_id'], bind['repo_id'],
+                                                bind['distributor_id'], options)
+            if report:
+                additional_tasks.extend(report.spawned_tasks)
+        except Exception, e:
+            errors.append(e)
+
+    error = None
+    if len(errors) > 0:
+        error = pulp_exceptions.PulpCodedException(error_codes.PLP0007, repo_id=repo_id)
+        error.child_exceptions = errors
+
+    return TaskResult(error=error, spawned_tasks=additional_tasks)
+
+
+def update_repo_and_plugins(repo, repo_delta, importer_config, distributor_configs):
+    """
+    Update a reposiory and its related collections.
+
+    All details do not need to be specified; if a piece is omitted it's configuration is not
+    touched, nor is it removed from the repository. The same holds true for the distributor_configs
+    dict, not every distributor must be represented.
+
+    This call will attempt to update the repository object, then the importer, then the
+    distributors. If an exception occurs during any of these steps, the updates stop and the
+    exception is immediately raised. Any updates that have already taken place are not rolled back.
+
+    Distributor updates are asynchronous as there could be a very large number of consumers to
+    update. Repository and importer updates are done synchronously.
+
+    :param repo: repository object
+    :type  repo: model.Repository
+    :param repo_delta: list of attributes to change and their new values; if None, no attempt to
+                       update the repository object will be made
+    :type  repo_delta: dict, None
+    :param importer_config: new configuration to use for the repo's importer; if None, no attempt
+                            will be made to update the importer
+    :type  importer_config: dict, None
+    :param distributor_configs: mapping of distributor ID to the new configuration to set for it
+    :type  distributor_configs: dict, None
+
+    :return: Task result that contains the updated repository object and additional spawned tasks
+    :rtype:  TaskResult
+
+    :raises pulp_exceptions.InvalidValue: if repo_delta is not a dictionary
+    """
+    if repo_delta:
+        if isinstance(repo_delta, dict):
+            repo.update_from_delta(repo_delta)
+            repo.save()
+        else:
+            raise pulp_exceptions.InvalidValue(['delta'])
```

Comment:
This is a new behavior (non dict would have been a 500), so I could raise a different error. `InvalidValue` is currently what the rest of the API and manger/controller code raises in this situation, what would you suggest?